### PR TITLE
Integrate xvm functionality

### LINKED ISSUE #84: Integrate XVM functionality to mainnet.
#### Description
Mainnet requires XVM functionality

#### Definition of "Done"
* We can invoke contracts on evm from wasm and vice versa.

#### Proposed solution:
* Copy [ink.rs](https://github.com/GoldenGateGGX/golden-gate/blob/main/runtime/testnet/src/ink.rs), [ethereum.rs](https://github.com/GoldenGateGGX/golden-gate/blob/main/runtime/testnet/src/ethereum.rs) to mainnet and integrate pallets to runtime
* Re-use [testnet](https://github.com/GoldenGateGGX/golden-gate/blob/main/node/src/service/testnet.rs) service for mainnet in node.

#### Useful links
* [Example how to test xvm](https://github.com/GoldenGateGGX/golden-gate/tree/main/examples/cross-vm-communication)
* [Transfer funds between Substrate and EVM address spaces](https://medium.com/astar-network/using-astar-network-account-between-substrate-and-evm-656643df22a0)

### PR DESCRIPTION
**Pull request by baidang201**
Closes #84

### CODE REVIEW

File: node/Cargo.toml [hunk 1]
```diff
@@ -106,7 +106,19 @@ testnet = [
 	"fp-storage",
 	"fc-storage",
 ]
-mainnet = ["golden-gate-runtime-mainnet"]
+mainnet = [
+	"golden-gate-runtime-mainnet",
+	"fc-cli",
```
Comment (akorchyn) [inline node/Cargo.toml:111]: probably can remove optionality, i dunno

File: runtime/mainnet/Cargo.toml [hunk 2]
```diff
@@ -67,10 +67,31 @@ pallet-treasury.workspace = true
...
+pallet-base-fee = { workspace = true }
+pallet-dynamic-fee = { workspace = true }
+pallet-ethereum = { workspace = true }
+pallet-evm = { workspace = true, features = ["forbid-evm-reentrancy"] }
+pallet-evm-chain-id = { workspace = true }
+pallet-hotfix-sufficients = { workspace = true }
+
 # GGX
 account-filter.workspace = true
 runtime-common.workspace = true
```
Comment (akorchyn) [inline runtime/mainnet/Cargo.toml:85]: runtime.common = {workspace = true, features = ["xvm"]}

File: runtime/mainnet/src/ethereum.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+			let account_id = T::AddressMapping::into_account_id(<Pallet<T>>::find_author());
+
+			let tip_commission = TREASURY_COMMISSION_FROM_TIP * tip.peek();
+			let (tip_commission, tip_after_commission) = tip.split(tip_commission);
+			OU::on_unbalanced(tip_commission);
+
+			let _ = C::deposit_into_existing(&account_id, tip_after_commission.peek());
+		}
+	}
+}
```
Comment (akorchyn) [inline runtime/mainnet/src/ethereum.rs:139]: probably good thing to move to currency manager to keep all thing in one place

**Comment by akorchyn** [comment]
It would be nice to add xvm test for your newly created method (prev PR)

File: runtime/mainnet/src/pos/currency.rs [hunk 4]
```diff
@@ -270,6 +283,128 @@ impl<T: Config> OnUnbalanced<NegativeImbalance<T>> for Author<T> {
...
+			// merge the imbalance caused by paying the fees and refunding parts of it again.
+			let adjusted_paid = paid.offset(refund_imbalance).same().unwrap_or_else(|_| {
+				<<T as pallet_evm::Config>::Currency as Currency<
+					<T as frame_system::Config>::AccountId,
+				>>::NegativeImbalance::zero()
+			});
+
+			let (base_fee, tip) = adjusted_paid.split(base_fee.unique_saturated_into());
+			// Handle base fee. Can be either burned, rationed, etc ...
+			T::FeeComissionRecipient::on_unbalanced(base_fee);
```
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/mainnet/src/pos/currency.rs:383]: sure

**Comment by baidang201** [comment]
> It would be nice to add xvm test for your newly created method (prev PR)
would be fine

File: runtime/mainnet/src/pos/currency.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/mainnet/src/pos/currency.rs]: fixed

File: node/Cargo.toml [hunk 5]
```diff
@@ -106,7 +106,18 @@ testnet = [
 	"fp-storage",
 	"fc-storage",
 ]
-mainnet = ["golden-gate-runtime-mainnet"]
+mainnet = [
+	"golden-gate-runtime-mainnet",
```
[the remark this answers is not in the document]
Comment (baidang201) [reply node/Cargo.toml:110]: golden-gate-runtime-testnet and golden-gate-runtime-mainnet  use the same libs?

File: examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml [hunk 6]
```diff
@@ -5,31 +5,23 @@ authors = ["[your_name] <[your_email]>"]
...
-    "ink_metadata/std",
-    "ink_env/std",
-    "ink_storage/std",
-    "ink_primitives/std",
+    "ink/std",
     "scale/std",
     "scale-info/std",
 ]
 ink-as-dependency = []
+e2e-tests = []
```
[the remark this answers is not in the document]
Comment (baidang201) [reply examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml:27]: ok, i will remove

File: examples/cross-vm-communication/evm-to-wasm/flipper/.gitignore
[the remark this answers is not in the document]
Comment (baidang201) [reply examples/cross-vm-communication/evm-to-wasm/flipper/.gitignore:1]: yes

File: node/src/rpc.rs [hunk 7]
```diff
@@ -5,14 +5,21 @@ use sc_consensus_manual_seal::rpc::{ManualSeal, ManualSealApiServer};
 use sc_rpc_api::DenyUnsafe;
 
 /// Full client dependencies.
-pub struct FullDeps<C, P, #[cfg(feature = "testnet")] A: sc_transaction_pool::ChainApi> {
+pub struct FullDeps<
+	C,
+	P,
+	#[cfg(feature = "mainnet")] A: sc_transaction_pool::ChainApi,
+	#[cfg(feature = "testnet")] A: sc_transaction_pool::ChainApi,
```
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/rpc.rs:12]: sure

[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/rpc.rs:12]: sure

File: node/src/rpc.rs [hunk 8]
```diff
@@ -5,14 +5,21 @@ use sc_consensus_manual_seal::rpc::{ManualSeal, ManualSealApiServer};
...
+> {
 	/// The client instance to use.
 	pub client: Arc<C>,
 	/// Transaction pool instance.
 	pub pool: Arc<P>,
 	/// Whether to deny unsafe calls
 	pub deny_unsafe: DenyUnsafe,
 
+	#[cfg(feature = "mainnet")]
```
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/rpc.rs:21]: sure

File: examples/cross-vm-communication/evm-to-wasm/flipper/.gitignore
[the remark this answers is not in the document]
Comment (baidang201) [reply examples/cross-vm-communication/evm-to-wasm/flipper/.gitignore:1]: fixed

File: examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml:27]: fixed

File: runtime/mainnet/src/pos/currency.rs [hunk 9]
```diff
@@ -270,6 +283,128 @@ impl<T: Config> OnUnbalanced<NegativeImbalance<T>> for Author<T> {
 	}
 }
 
+impl<T> OnChargeEVMTransaction<T> for Pallet<T>
```
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/mainnet/src/pos/currency.rs:291]: fixed

File: node/src/rpc.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/rpc.rs]: fixed

File: examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply examples/cross-vm-communication/evm-to-wasm/flipper/Cargo.toml]: fixed

File: node/src/rpc.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/rpc.rs:16]: don’t need to fix it in your PR

https://doc.rust-lang.org/cargo/reference/features.html#feature-unification

[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/src/rpc.rs:16]: We don't need to fix this here (it was already broken).

Ticketed: https://github.com/GoldenGateGGX/golden-gate/issues/97#issue-1710445624

File: node/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply node/Cargo.toml:110]: That's right.

We don't need to fix this here (as noted, it was already broken before).

Ticketed: https://github.com/GoldenGateGGX/golden-gate/issues/97#issue-1710445624

**Review by xcthulhu — APPROVED** [review]

**Pull request merged**